scipy/optimize/minpack.py

```diff
@@ -510,28 +523,42 @@ def curve_fit(f, xdata, ydata, p0=None, sigma=None, absolute_sigma=False,
         values will all be 1 (if the number of parameters for the function
         can be determined using introspection, otherwise a ValueError
         is raised).
-    sigma : None or M-length sequence, optional
-        If not None, the uncertainties in the ydata array. These are used as
-        weights in the least-squares problem
-        i.e. minimising ``np.sum( ((f(xdata, *popt) - ydata) / sigma)**2 )``
-        If None, the uncertainties are assumed to be 1.
+    sigma : None or M-length sequence or MxM array, optional
+        Determines the errors in `ydata`. If we define residuals as
+        ``r = ydata - f(xdata, *popt)``, then the interpretation of `sigma`
+        depends on its number of dimensions:
+
+            - A 1-d `sigma` should contain values of standard deviations of
+              errors in `ydata`. In this case, the optimized function is
+              ``chisq = sum((r / sigma) ** 2)``.
+
+            - A 2-d `sigma` should contain the covariance matrix of
+              errors in `ydata`. In this case, the optimized function is
+              ``chisq = r.T @ inv(sigma) @ r``.
+
+              .. versionadded:: 0.19
+
+        None (default) is equivalent of 1-d `sigma` filled with ones.
     absolute_sigma : bool, optional
-        If False, `sigma` denotes relative weights of the data points.
-        The returned covariance matrix `pcov` is based on *estimated*
-        errors in the data, and is not affected by the overall
-        magnitude of the values in `sigma`. Only the relative
-        magnitudes of the `sigma` values matter.
-
-        If True, `sigma` describes one standard deviation errors of
-        the input data points. The estimated covariance in `pcov` is
-        based on these values.
+        If True, `sigma` is used in an absolute sense and the estimated parameter
+        covariance `pcov` reflects these absolute values.
+
+        If False, only the relative magnitudes of the `sigma` values matter.
+        The returned parameter covariance matrix `pcov` is based on scaling
+        `sigma` by a constant factor. This constant is set by demanding that the
+        reduced `chisq` for the optimal parameters `popt` when using the
+        *scaled* `sigma` equals unity. In other words, `sigma` is scaled to
+        match the sample variance of the residuals after the fit. In this case,
```

Comment:
"In this case, `pcov` is not affected by the overall magnitude of the values in `sigma`." --- this sentence is unnecessary in my opinion (if we compare it with the starting one). Sorry I didn't say it explicitly before.

Feel free to leave it if you disagree.